File: project_euler/problem_202/solution.py
"""
Problem 202: Laserbeam

A laser beam enters vertex C of an equilateral triangle with mirrored sides,
reflects exactly 12017639147 times off the internal surfaces, and exits
through vertex C. Count the number of distinct beam paths.

Approach:
- Unfold reflections into a triangular tessellation.
- n = (R + 3) / 2 = 6008819575 is the lattice parameter.
- Count b in [1, n-1] with b congruent to 2 modulo 3 and gcd(b, n) = 1.
- Use Mobius inversion over squarefree divisors of n.

n = 6008819575 = 5^2 * 11 * 17 * 23 * 29 * 41 * 47
"""


def solve():
    r = 12017639147
    n = (r + 3) // 2

    primes = []
    tmp = n
    p = 2
    while p * p <= tmp:
        if tmp % p == 0:
            primes.append(p)
            while tmp % p == 0:
                tmp //= p
        p += 1
    if tmp > 1:
        primes.append(tmp)

    answer = 0
    for mask in range(1 << len(primes)):
        d = 1
        bits = 0
        for i, prime in enumerate(primes):
            if mask & (1 << i):
                d *= prime
                bits += 1

        mu = 1 if bits % 2 == 0 else -1
        nd = n // d

        if d % 3 == 0:
            count = 0
        else:
            inverse_mod_3 = 1 if d % 3 == 1 else 2
            residue = (2 * inverse_mod_3) % 3
            upper = nd - 1
            if residue == 0:
                count = (upper - 3) // 3 + 1 if upper >= 3 else 0
            else:
                count = (upper - residue) // 3 + 1 if upper >= residue else 0

        answer += mu * count

    print(answer)


if __name__ == "__main__":
    solve()

File: project_euler/problem_202/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    long long R = 12017639147LL;
    long long n = (R + 3) / 2;

    vector<long long> primes;
    long long tmp = n;
    for (long long p = 2; p * p <= tmp; p++) {
        if (tmp % p == 0) {
            primes.push_back(p);
            while (tmp % p == 0) {
                tmp /= p;
            }
        }
    }
    if (tmp > 1) {
        primes.push_back(tmp);
    }

    long long answer = 0;
    for (int mask = 0; mask < (1 << static_cast<int>(primes.size())); mask++) {
        long long d = 1;
        int bits = 0;
        for (int i = 0; i < static_cast<int>(primes.size()); i++) {
            if (mask & (1 << i)) {
                d *= primes[i];
                bits++;
            }
        }

        long long mu = (bits % 2 == 0) ? 1 : -1;
        long long nd = n / d;
        long long count = 0;

        if (d % 3 != 0) {
            long long inverseMod3 = (d % 3 == 1) ? 1 : 2;
            long long residue = (2 * inverseMod3) % 3;
            long long upper = nd - 1;

            if (residue == 0) {
                if (upper >= 3) {
                    count = (upper - 3) / 3 + 1;
                }
            } else {
                if (upper >= residue) {
                    count = (upper - residue) / 3 + 1;
                }
            }
        }

        answer += mu * count;
    }

    cout << answer << '\n';
    return 0;
}
